File: 1001~1500/1392. Longest Happy Prefix/main.py
"""
Source: https://leetcode.com/problems/longest-happy-prefix/
Date: 2023/3/4
Skill:
Ref:
Runtime: 295 ms, faster than 68.73%
Memory Usage: 19.1 MB, less than 21.81%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def longestPrefix(self, s: str) -> str:
        dp = [0 for _ in range(len(s))]
        for i in range(1, len(s)):
            j = dp[i-1]
            while j > 0 and s[j] != s[i]:
                j = dp[j-1]
            dp[i] = j + (s[j] == s[i])

        return s[:dp[len(s)-1]]


if __name__ == "__main__":
    s = Solution()

File: 1001~1500/1392. Longest Happy Prefix/main.cpp
/**
 * Source: https://rb.gy/52ch0
 * Date: 2023/9/10
 * Skill:
 * Runtime: WA
 * Memory Usage:
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long

typedef pair<int, int> pairs;

typedef unsigned long long ULL;
typedef pair<ULL, ULL> PULL;


class Solution {
public:
    string longestPrefix(string s) {
        int sz = s.length(), res = 0;
        ULL hash1 = 0, hash2 = 0, base = 26, power = 1;
        for (int i = 0; i < sz - 1; i++) {
            hash1 = hash1 * base + s[i] - 'a';
            hash2 += (s[sz - 1 - i] - 'a') * power;
            if (hash1 == hash2)
                res = i + 1;
            power *= base;
        }
        return s.substr(0, res);
    }
};


static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

int main() {
    Solution s;
    string ss = "level";
    string res = s.longestPrefix(ss);
    cout << res << endl;
}